### PR TITLE
Support for optional/required properties on nested object parameters

### PR DESCRIPTION
**Pull request by filmaj**
###  Summary

This is related to HERMES-4071.

I feel like working on this has aged me by like 3 years.

Summary of user-facing changes:

- Recommend wrapping any typed-object parameters in a new `DefineProperty` helper function

Summary of more internal changes:

- Creating `ParameterDefinitionWithGenerics` as a generic version of `ParameterDefinition`
- Passing generics to `TypedObjectParameterDefinition` so that a constraint exists between `properties` and `required`
  - Creating `TypedObjectParameter` as a way to pass around the generic-less schema/shape for reference in `extends` constraints
- Move from using TS `type` to `interface` so that JSDoc comments are inherited appropriately

# TODO

- [x] `DefineParameter` name OK, or something else?

### CODE REVIEW

File: src/dev_deps.ts
Comment (shapirone) [inline src/dev_deps.ts:31]: I'm thinking we might want to include these in our `src/test_utils.ts` folder instead

Comment (filmaj) [reply src/dev_deps.ts:31]: It's been a while - what's your opinion on this now @shapirone ?

File: src/functions/types.ts
Comment (filmaj) [inline src/functions/types.ts:84]: This particular change is just grouping all the types that yield a `string`

File: src/dev_deps.ts
Comment (shapirone) [reply src/dev_deps.ts:31]: I feel the same, `dev_deps` is intended for importing so AFAIK we shouldn't be defining our own things in here 😅

File: src/functions/types.ts
Comment (filmaj) [inline src/functions/types.ts:91]: Same here for `number`

File: src/functions/types.ts
Comment (filmaj) [inline src/functions/types.ts:112]: This is the main change enabling proper object sub-property optionality autocomplete in a function implementation context.

File: src/parameters/define_parameter.ts
Comment (filmaj) [inline src/parameters/define_parameter.ts:7]: The only purpose this method serves is to create a 'hook' for the type system to be able to apply the two generics below this line, and in particular, that is important because the generics sequentially rely on each other; that is, the third generic relies on the previous two named generics, and the second generic relies on the first named generic.

File: src/parameters/define_parameter_test.ts
Comment (filmaj) [inline src/parameters/define_parameter_test.ts:23]: Really wish we could solve this issue 😢

File: src/parameters/mod.ts
Comment (filmaj) [inline src/parameters/mod.ts:49]: It's a complete mystery to me why previously, with no recursion guard whatsoever, the old code didn't lead to an infinite-recursion TS error.

File: src/schema/slack/functions/open_form.ts
Comment (filmaj) [inline src/schema/slack/functions/open_form.ts:30]: Huh, should these be removed?

File: src/schema/slack/types/custom/form_input.ts
Comment (filmaj) [inline src/schema/slack/types/custom/form_input.ts:5]: I think _technically_ we can avoid using `DefineParameter` here as I believe the call to `DefineType` below should function in the same way in terms of providing an entrypoint for the props/required props generics to come into play.

That said, this is all internal stuff not directly exposed to the dev, and I'd rather we use `DefineParameter` to wrap all typed object definitions whenever we can, even if redundant.

Open to either. This same comment applies to all three of the Slack custom types.

File: src/functions/tester/mod.ts
Comment (shapirone) [reply src/functions/tester/mod.ts:45]: The `SlackFunctionTester` needs to be in `src` so that it's shared with developers, no? It's what allows them to run unit tests on their functions.

File: src/functions/types.ts
Comment (shapirone) [reply src/functions/types.ts:56]: Ah yeah, might be good to move recursion guard logic to a shared space then!

File: src/functions/tester/mod.ts
Comment (filmaj) [reply src/functions/tester/mod.ts:45]: Damn I keep forgetting that this is exposed to devs! Good call, resolving.

File: src/functions/types.ts
Comment (filmaj) [reply src/functions/types.ts:56]: Hmm, maybe `src/types.ts`? `src/shared_types.ts`?

File: src/parameters/define_parameter_test.ts
Comment (shapirone) [reply src/parameters/define_parameter_test.ts:23]: Agreed, the solution we found leads to another inexplicable issue where devs may need to instantiate a variable that uses `DefineParameter` outside of a `DefineFunction` call, which seems unacceptable to me

File: src/schema/slack/types/custom/form_input.ts
Comment (shapirone) [reply src/schema/slack/types/custom/form_input.ts:5]: Good catch, I'd like for us to just use `DefineType` rather than spreading the result of `DefineParameter` into a `DefineType` call, since I think that's what we would recommend to devs.

Comment (filmaj) [reply src/schema/slack/types/custom/form_input.ts]: K I updated this in https://github.com/slackapi/deno-slack-sdk/pull/124/commits/26345ff5ce097f1c06efd90fa96dd9713933bd2d and also added a test for the internal `FormInput` slack type to make sure it behaves as we would expect in a function runtime context.

File: src/schema/slack/functions/open_form.ts
Comment (WilliamBergamin) [reply src/schema/slack/functions/open_form.ts:30]: This will eventually be added back when the script is ran again, but these are generated changes

Comment (shapirone) [reply src/schema/slack/functions/open_form.ts:30]: Do we know why or how these made it into the PR? Was it me? 😅

Comment (WilliamBergamin) [reply src/schema/slack/functions/open_form.ts:30]: Changing the contents of the `function.json` could easily have caused a change in the generated files

Comment (shapirone) [reply src/schema/slack/functions/open_form.ts:30]: But someone would have had to run the generation script to change these files. IMO we shouldn't commit these changes

Comment (filmaj) [reply src/schema/slack/functions/open_form.ts:30]: It was hard for me to figure out where this blame came from haha. Should I just put these descriptions back ?

File: src/functions/types.ts
Comment (shapirone) [reply src/functions/types.ts:56]: Something like `src/type_utils.ts` would match `src/test_utils.ts` 🤷

File: src/manifest/manifest_schema.ts
Comment (shapirone) [inline src/manifest/manifest_schema.ts:3]: ultra nit: The fact that we didn't move this to `parameters/types.ts` as well sticks out to me here

File: src/manifest/manifest_test.ts
Comment (shapirone) [inline src/manifest/manifest_test.ts:35]: All changes in this file seem unnecessary, my best guess is that they're remnants of when we didn't support `required` being `undefined` (w/ some formatting changes)

File: src/functions/types.ts
Comment (filmaj) [reply src/functions/types.ts]: Aight done in fb96d03

File: src/parameters/define_parameter_test.ts
Comment (shapirone) [inline src/parameters/define_parameter_test.ts:1]: Need to rename the file we're importing from, and this test file

File: src/parameters/mod.ts
Comment (shapirone) [inline src/parameters/mod.ts:87]: Nit: We could do what we did for runtime functions and remove the need to specify `ObjectParameterPropertyTypes<Def["properties"]>` twice

```suggestion
type ObjectParameterVariableType<
  Def extends TypedObjectParameter,
> =
  & ObjectParameterPropertyTypes<Def["properties"]>
  & (Def["additionalProperties"] extends false ? Record<never, never>
    : {
      // deno-lint-ignore no-explicit-any
      [key: string]: any;
    });
```

File: src/parameters/mod.ts
Comment (shapirone) [inline src/parameters/mod.ts:121]: Can we do anything for this todo?

File: src/parameters/mod.ts
Comment (shapirone) [inline src/parameters/mod.ts:103]: 💭 Early on in our exploration we moved to this `switch` statement. Do we think that using a switch statement is still preferable considering we needed to add an if statement above?

File: src/manifest/manifest_schema.ts
Comment (filmaj) [reply src/manifest/manifest_schema.ts]: Done in 54e5989

File: src/manifest/manifest_test.ts
Comment (filmaj) [reply src/manifest/manifest_test.ts:35]: Fixed in d09eb32

File: src/parameters/define_parameter_test.ts
Comment (filmaj) [reply src/parameters/define_parameter_test.ts]: Fixed in 9b73d87

File: src/parameters/mod.ts
Comment (filmaj) [reply src/parameters/mod.ts:59]: Can we chalk it up as something to tackle later and not block this PR anymore? I can't fully remember what the coment is about to be honest

File: src/schema/slack/functions/open_form.ts
Comment (shapirone) [reply src/schema/slack/functions/open_form.ts:30]: Yes please, none of our built-ins should have changes in this PR

File: src/parameters/mod.ts
Comment (filmaj) [reply src/parameters/mod.ts:41]: Probably doesn't make any sense anymore. Should I remove it and flip back to `if`s?

Comment (shapirone) [reply src/parameters/mod.ts:41]: Yeah that makes sense to me

File: src/schema/slack/types/custom/form_input.ts
Comment (shapirone) [inline src/schema/slack/types/custom/form_input.ts:22]: I think both of these are required, but we may need to double check

File: src/schema/slack/functions/open_form.ts
Comment (filmaj) [reply src/schema/slack/functions/open_form.ts]: Done in bdbdce8

File: src/schema/slack/types/custom/form_input.ts
Comment (filmaj) [reply src/schema/slack/types/custom/form_input.ts:22]: How/where can I double check?

File: src/parameters/mod.ts
Comment (shapirone) [reply src/parameters/mod.ts:57]: Heh, I also didn't follow what the comment was concerned about which is why I was hoping we could resolve before merging 😅 Fixing it by clarifying or removing isn't a PR blocker but would be nice to resolve rather than leaving it in here forever without clarifying

Comment (filmaj) [reply src/parameters/mod.ts]: I removed it in 30d8fef - the initializing variable at the top of this method initially assigns `null` to it but with the way the logic works out now, it will never remain `null`, so I think it is safe to remove.

File: src/schema/slack/types/custom/form_input.ts
Comment (shapirone) [reply src/schema/slack/types/custom/form_input.ts]: I checked by testing an app that uses `OpenForm` (which uses `FormInput`) and commented out both `elements` and `required`. Turns out `elements` is required, but `required` is optional! Updated accordingly

**Review by shapirone — APPROVED** [review]
lgtm

**Pull request merged**
